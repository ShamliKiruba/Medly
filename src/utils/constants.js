export const LOAN_DETAILS = [
    {
        label: 'Loan EMI',
        key: 'emi',
    },
    {
        label: 'Total Interest Payable',
        key: 'totalInterest',
    },
    {
        label: 'Total Payment',
        key: 'totalPayment',
    }
];

export const MONTH_LIST = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];

export const TABLE_TITLE = ['Year', 'Principal (A)', 'Interest (B)', 'Total Payment (A + B)', 'Balance', 'Loan Pain To Date'];

export const INIT_TABLEDATA = {
    thead: TABLE_TITLE,
    tbody: [{
        year: new Date().getFullYear(),
        month: '',
        principal: 1,
        interest: 1,
        total: 1,
        balance: 1,
        loanPaid: 1
    }],
};

export const INIT_DATE = {
    date: (new Date().getDate()) < 10 ? `0${(new Date().getDate())}`: `${new Date().getDate()}`,
    month: (new Date().getMonth() + 1) < 10 ? `0${(new Date().getMonth() + 1)}`: `${new Date().getMonth() + 1}`,
    year: `${new Date().getFullYear()}`,
};

export const INIT_LOAN_INFO = {
    loanAmount: 5000000,
    interestRate: 10.5,
    loanTenure: 20
};

export const INIT_ANNUAL_INFO = {
    emi: 0,
    totalPayment: 0,
    totalInterest: 0,
};