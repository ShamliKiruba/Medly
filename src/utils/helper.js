import { MONTH_LIST, TABLE_TITLE } from './constants';

export const range = (limit, step, spacing, unit) => {
    let marks = [];
    for(let i = 0; i <= limit; i = i+step) {
        let obj = {
            value: i,
            label: unit ? (i+unit) : i,
            spacing
        }
        marks.push(obj);
    }
    return marks;
};

export const LOAN_AMOUNT_MARKER = range(200, 25, 60, 'L');
export const INTEREST_RATE_MARKER = range(20, 2.5, 60);
export const TENURE_YEAR_MARKER = range(30, 5, 80);
export const TENURE_MONTH_MARKER = range(360, 60, 80);

export const TENURE_MONTH = {
    label: 'Loan Tenure',
    id:'loanTenure',
    defaultValue: 20,
    step: 6,
    min: 0,
    max: 360,
    marks: TENURE_MONTH_MARKER,
    selectedUnit: 'Mon',
    units: ['Yr', 'Mon']
};

export const TENURE_YEAR = {
    label: 'Loan Tenure',
    id:'loanTenure',
    defaultValue: 20,
    step: 0.5,
    min: 0,
    max: 30,
    selectedUnit: 'Yr',
    marks: TENURE_YEAR_MARKER,
    units: ['Yr', 'Mon']
};

export const RANGE_SLIDER = [
    {
        label: 'Home Loan Amount',
        id: 'loanAmount',
        defaultValue: 25,
        step: 50000,
        min: 0,
        max: 20000000,
        marks: LOAN_AMOUNT_MARKER,
        units: ['₹']
    },
    {
        label: 'Interest Rate',
        id:'interestRate',
        defaultValue: 10.5,
        step: 0.5,
        min: 0,
        max: 20,
        marks: INTEREST_RATE_MARKER,
        units: ['%']
    },
    TENURE_YEAR,
];

export const calculateEMI = (loanInfo, duration) => {
    let interest = ((loanInfo.interestRate /12) * 0.01);
    let timeDuration = duration === 'Yr' ? (loanInfo.loanTenure * 12) : (loanInfo.loanTenure)
    let a = Math.pow((1 + interest), timeDuration)
    let b = a - 1;
    let emi = loanInfo.loanAmount * (interest) * (a/b);
    let totalPayment =  emi * timeDuration;
    let totalInterest = (totalPayment - loanInfo.loanAmount);
    return {
        emi: Math.round(emi),
        totalPayment: Math.round(totalPayment),
        totalInterest: Math.round(totalInterest),
    }
};

export const createTable = (props, dateObj) => {
    let durationInMonths = props.duration === 'Yr' ? props.loanInfo.loanTenure * 12 : props.loanInfo.loanTenure;
    let sumA = 0, sumB = 0, sumC = 0, sumD = 0, sumE = 0, parent = {}, arr = [], parentArr = [], year = Number(dateObj.year);
    let emi = Number(props.annualInfo.emi);
    let month = Number(dateObj.month);
    let balanceAmount = props.loanInfo.loanAmount, sumOfPrincipal = 0;
    let initialAmount = props.loanInfo.loanAmount;
    let childAccumulator = [];
    for(let i=0; i<durationInMonths;i++) {
        let temp = {
            year: year,
            month: MONTH_LIST[month-1],
            total: emi,
            loanPaid:  0
        };
        let interestComponent = (props.loanInfo.interestRate * balanceAmount)/1200;
        let principalComponent = emi - interestComponent;
        temp.balance = balanceAmount - principalComponent; // final
        temp.interest = interestComponent;
        temp.principal = principalComponent;

        balanceAmount = temp.balance;
        sumOfPrincipal+= temp.principal;
        temp.loanPaid = ((sumOfPrincipal / initialAmount) * 100).toFixed(2);

        sumA+=temp.principal;
        sumB+=temp.interest;
        sumC+=temp.total;
        
        childAccumulator.push({
            month: temp.month,
            principal: `₹ ${convertValue(Math.round(temp.principal))}`,
            interest: `₹ ${convertValue(Math.round(temp.interest))}`,
            total: `₹ ${convertValue(emi)}`,
            balance: `₹ ${convertValue(Math.round(temp.balance))}`,
            loanPaid: `${temp.loanPaid} %`,
            year: temp.year,
        });
        if(MONTH_LIST[month-1] === 'Dec'  || (i === durationInMonths - 1)) {
            sumD=balanceAmount;
            sumE+=Number(temp.loanPaid);
            parent = {
                year, 
                sumA: `₹ ${convertValue(Math.round(sumA))}`, 
                sumB: `₹ ${convertValue(Math.round(sumB))}`, 
                sumD: (i === durationInMonths - 1) ? 0 : `₹ ${convertValue(Math.round(sumD))}`, 
                sumC: `₹ ${convertValue(sumC)}`,
                sumE: `${sumE} %`
            };
            arr.push(parent, ...childAccumulator);
            parentArr.push({
                year, 
                sumA: Math.round(sumA), 
                sumB: Math.round(sumB), 
                sumD: (i === durationInMonths - 1) ? 0 : Math.round(sumD), 
                sumC, 
                sumE
            });
            year++;
            sumA = 0; sumB = 0; sumC = 0; sumE = 0;
            month = 1;
            childAccumulator = [];
        } else {
            month++;
        }
    }
    return { tableData: {
        thead: TABLE_TITLE,
        tbody: arr
    }, parentArr };
};

export const convertValue = (value = 0) => {
    return value.toLocaleString('en-IN');
}