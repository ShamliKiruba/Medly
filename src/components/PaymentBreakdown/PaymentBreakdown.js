import React, { useState, useEffect, useCallback } from 'react';
import { createTable } from '../../utils/helper';
import { updateBarChart } from '../common/Chart';
import { INIT_TABLEDATA, INIT_DATE } from '../../utils/constants';
import DatePicker from '../common/DatePicker/DatePicker';
import Accordion from '../common/Accordion/Accordion';

const PaymentBreakdown = (props) => {
    const [tableData, setTableData] = useState(INIT_TABLEDATA);
    const [selectedYear, updateSelectedYear] = useState([]);
    const [dateObj, setPeriod] = useState(INIT_DATE);
    
    const createTableData = useCallback(() => {
        const { tableData, parentArr } = createTable(props, dateObj);
        setTableData(tableData);
        updateBarChart(parentArr);
    },[props, dateObj]);


    useEffect(() => {
        createTableData();
    }, [createTableData, props.annualInfo.emi]);


    const showSubTable = (obj) => {
        let year = obj.year;
        let yearList = [...selectedYear];
        if (!yearList.includes(year)) {
            //expand
            yearList.push(year);
        } else {
            //collapse
            yearList.splice(yearList.indexOf(year), 1);
        }
            
        updateSelectedYear(yearList);
    };

    
    const updatePeriod = (value) => {
        let obj = value.split('-');
        let temp = dateObj;
        temp.year = obj[0];
        temp.month = obj[1];
        temp.date = obj[2];
        setPeriod(prevState => ({
            ...prevState,
            temp,
        }));
        createTableData();
    };

    return (
        <div className="part-two">
            <div className="date" id="date">
                <p>Schedule showing EMI payments starting from</p>
                <DatePicker 
                    value={`${dateObj.year}-${dateObj.month}-${dateObj.date}`}
                    onChange = {(value) => {updatePeriod(value)}}
                />
            </div>
            <div id="container">

            </div>
            <div className="table-data">
                <Accordion 
                    tableData={tableData}
                    parentKey='sumA'
                    commonKey='year'
                    selectedRow={selectedYear}
                    showSubTable={showSubTable}
                />
            </div>
        </div>
    )
}

export default PaymentBreakdown;
