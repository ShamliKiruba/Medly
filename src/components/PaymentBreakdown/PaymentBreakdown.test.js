import React from 'react';
import { fireEvent,render } from '@testing-library/react';
import PaymentBreakdown from './PaymentBreakdown';
import { createTable } from '../../utils/helper'

const props = {
	annualInfo:{
		emi: 0,
		totalPayment: 0,
		totalInterest: 0,
	},
	duration:'Yr',
	loanInfo:{
		loanAmount: 5000000,
		interestRate: 10.5,
		loanTenure: 20
	}
}
const dateObj = {
	date: '20',
	month: '06',
	year: '2022'
};
describe('Payment breakdown', () => {
	it('renders date picker', () => {
		const { getByText } = render(<PaymentBreakdown 
		{...props}
		/>);
		const header = getByText('Schedule showing EMI payments starting from');
		expect(header).toBeInTheDocument();
		expect(header).toMatchSnapshot();
	});
	
	it('should create table on date change', () => {
		const { container } = render(<PaymentBreakdown 
			{...props}
		/>);
		fireEvent.change(container.querySelector('input[type=date]'), {
			target : { value : dateObj}
		});
		const mock = createTable(props, dateObj);
		expect(mock).toHaveProperty('parentArr');
	});
	
	it('should show sub table', () => {
		const { container } = render(<PaymentBreakdown 
			{...props}
		/>);
		fireEvent.click(container.querySelector('.expand-collapse'));
		const childClass = container.querySelector('tr:nth-child(2)').classList;
		expect(childClass).toContain('show');
	});
});
