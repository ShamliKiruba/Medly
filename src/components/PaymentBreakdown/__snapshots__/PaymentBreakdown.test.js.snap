// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Payment breakdown renders date picker 1`] = `
<p>
  Schedule showing EMI payments starting from
</p>
`;
